File: src/com/medium/HouseRobberII_213.java
/**
 * 
 */
package com.medium;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class HouseRobberII_213 {
    public static void main(String[] args) throws Exception {
        HouseRobberII_213 robber = new HouseRobberII_213();
        System.out.println(robber.rob(new int[] { 1, 2, 3, 4, 5, 1, 2, 3, 4, 5 }));
    }

    // consider 2 case:
    // case 1: rob 0 => rob(0:N-1)
    // case 2: not rob 0 => rob(1:N)
    public int rob(int[] nums) {
        if (nums.length == 0 || nums == null) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        // System.out.println(rob(nums, 0, nums.length - 2));
        // System.out.println(rob(nums, 1, nums.length - 1));
        return Math.max(rob2(nums, 0, nums.length - 2), rob2(nums, 1, nums.length - 1));
    }

    // Use DP with memorize
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 34 MB, less than 100.00%
    private int rob(int[] nums, int mIndexStart, int mIndexEnd) {
        int mLength = mIndexEnd - mIndexStart + 1;
        if (mLength == 0) {
            return 0;
        }
        if (mLength == 1) {
            return nums[mIndexStart];
        }
        int[] mDP = new int[mLength];
        mDP[0] = nums[mIndexStart];
        mDP[1] = Math.max(nums[mIndexStart], nums[mIndexStart + 1]);
        for (int i = 2; i < mLength; i++) {
            mDP[i] = Math.max(mDP[i - 2] + nums[mIndexStart + i], mDP[i - 1]);
            // System.out.printf("mDP[%d]:%d\n", i, mDP[i]);
        }
        return mDP[mLength - 1];
    }

    // Use DP without memorize
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 34.2 MB, less than 100.00%
    private int rob2(int[] nums, int mIndexStart, int mIndexEnd) {
        int mPrevious = 0, mCurrent = 0;
        for (int i = mIndexStart; i <= mIndexEnd; i++) {
            int mTmp = mCurrent;
            mCurrent = Math.max(mPrevious + nums[i], mCurrent);
            mPrevious = mTmp;
        }
        return mCurrent;
    }
}

File: Golang_src/medium/HouseRobberII_213.go
package medium

// Use DP with memorize:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.1 MB, less than 45.16%
func rob(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return 0
	}
	length := len(nums)
	if length == 1 {
		return nums[0]
	}

	return max(subRob(nums, 0, length-2), subRob(nums, 1, length-1))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func subRob(nums []int, indexStart, indexEnd int) int {
	length := indexEnd - indexStart + 1

	if length == 0 {
		return 0
	}
	if length == 1 {
		return nums[indexStart]
	}
	dp := make([]int, length)
	dp[0] = nums[indexStart]
	dp[1] = max(nums[indexStart], nums[indexStart+1])
	for i := 2; i < length; i++ {
		dp[i] = max(dp[i-1], dp[i-2]+nums[indexStart+i])
	}

	return dp[length-1]
}

// Use DP without memorize:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 45.16%
func rob2(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return 0
	}
	length := len(nums)
	if length == 1 {
		return nums[0]
	}

	return max(subRob2(nums, 0, length-2), subRob2(nums, 1, length-1))
}

func subRob2(nums []int, indexStart, indexEnd int) int {
	rob, notRob := 0, 0

	for i := indexStart; i <= indexEnd; i++ {
		tmp := rob
		rob = max(rob, notRob+nums[i])
		notRob = tmp
	}
	return rob
}
